kernel/src/loader/mod.rs

```diff
@@ -134,27 +138,37 @@ where
             .seek(SeekFrom::Start(phdr.p_offset))
             .map_err(|_| Error::SeekKernelStart)?;
 
-        let mem_offset = GuestAddress(phdr.p_paddr as usize);
-        if mem_offset.offset() < start_address {
+        if (phdr.p_paddr as usize) < start_address {
             return Err(Error::InvalidProgramHeaderAddress);
         }
 
-        guest_mem
-            .read_to_memory(mem_offset, kernel_image, phdr.p_filesz as usize)
-            .map_err(|_| Error::ReadKernelImage)?;
+        read_to_memory(phdr.p_paddr as usize, kernel_image, phdr.p_filesz as usize)?;
     }
 
-    Ok(GuestAddress(ehdr.e_entry as usize))
+    Ok(ehdr.e_entry as usize)
 }
 
 #[cfg(target_arch = "aarch64")]
-pub fn load_kernel<F>(
-    guest_mem: &GuestMemory,
-    kernel_image: &mut F,
+/// Loads a kernel from a vmlinux image using a given callback.
+///
+/// # Arguments
+///
+/// * `guest_mem` - The guest memory region the kernel is written to.
+/// * `kernel_image` - Input vmlinux image.
+/// * `start_address` - Kernel start address in the guest memory.
+/// * `read_to_memory` - Closure to write the contents of the image.
```

Comment:
Nit: maybe update the callback name to write.

It would also be nice to write a very short description of the callback arguments. Now that the method on the GuestMemory is not called directly it is a bit harder to follow the code and understand how to use the load_kernel method.